src/test/java/uk/gov/hmcts/reform/iacaseapi/domain/handlers/presubmit/AppealReferenceNumberHandlerTest.java

```diff
@@ -0,0 +1,137 @@
+                .isExactlyInstanceOf(IllegalStateException.class);
+    }
+
+    @Test
+    public void should_get_next_appeal_reference_number() {
+
+        String protectionAppealType = "protection";
+
+        when(appealReferenceNumberGenerator.getNextAppealReferenceNumberFor(protectionAppealType))
+                .thenReturn(Optional.of("the-next-appeal-reference-number"));
+
+        PreSubmitCallbackResponse<AsylumCase> submitCallbackResponse =
+                underTest.handle(ABOUT_TO_SUBMIT, submitAppealCallbackForProtectionAsylumCase());
+
+        assertThat(submitCallbackResponse.getData().getAppealReferenceNumber().get())
+                .isEqualTo("the-next-appeal-reference-number");
+    }
+
+    @Test
+    public void doesnt_get_next_appeal_reference_number_if_already_present() {
+
+        Optional<String> appealReference = Optional.of("some-unexpected-reference-number");
+
+        when(callback.getCaseDetails()).thenReturn(caseDetails);
+        when(caseDetails.getCaseData()).thenReturn(caseData);
+        when(caseData.getAppealReferenceNumber()).thenReturn(appealReference);
+        when(callback.getEvent()).thenReturn(Event.SUBMIT_APPEAL);
+
+        PreSubmitCallbackResponse preSubmitCallbackResponse = underTest.handle(ABOUT_TO_SUBMIT, callback);
+
+        verifyZeroInteractions(appealReferenceNumberGenerator);
+
+        assertThat(caseData.getAppealReferenceNumber()).isSameAs(appealReference);
+
+        assertThat(preSubmitCallbackResponse.getErrors())
+                .containsExactly("Sorry, there was a problem submitting your appeal case");
+    }
+
+    @Test
+    public void should_not_allow_null_arguments() {
+
+        assertThatThrownBy(() -> underTest.canHandle(null, callback))
+                .hasMessage("callbackStage must not be null")
+                .isExactlyInstanceOf(NullPointerException.class);
+
+        assertThatThrownBy(() -> underTest.canHandle(PreSubmitCallbackStage.ABOUT_TO_SUBMIT, null))
+                .hasMessage("callback must not be null")
+                .isExactlyInstanceOf(NullPointerException.class);
+
+        assertThatThrownBy(() -> underTest.handle(null, callback))
+                .hasMessage("callbackStage must not be null")
+                .isExactlyInstanceOf(NullPointerException.class);
+
+        assertThatThrownBy(() -> underTest.handle(PreSubmitCallbackStage.ABOUT_TO_SUBMIT, null))
+                .hasMessage("callback must not be null")
+                .isExactlyInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    public void should_add_error_to_callback_response_if_unable_to_generate_appeal_reference() {
+
+        when(appealReferenceNumberGenerator.getNextAppealReferenceNumberFor("protection"))
+                .thenReturn(Optional.empty());
+
+        PreSubmitCallbackResponse<AsylumCase> submitCallbackResponse =
+                underTest.handle(ABOUT_TO_SUBMIT, submitAppealCallbackForProtectionAsylumCase());
```

Comment:
What is the text fixture class giving us that a mock wouldn't?
A unit test should be isolated from the rest of the system as far as reasonable, and it creates real classes.